### ISSUE TITLE
Nodesource SNI causing build failure on CentOS 6.x

### ISSUE THREAD

**Issue by geerlingguy** [geerlingguy/ansible-role-nodejs#46]
See error message:

```
TASK [role_under_test : Import Nodesource RPM key.] ****************************
fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "msg": "Failed to validate the SSL certificate for rpm.nodesource.com:443. Make sure your managed systems have a valid CA certificate installed. If the website serving the url uses SNI you need python >= 2.7.9 on your managed machine or you can install the `urllib3`, `pyopenssl`, `ndg-httpsclient`, and `pyasn1` python modules to perform SNI verification in python >= 2.6. You can use validate_certs=False if you do not need to confirm the servers identity but this is unsafe and not recommended. Paths checked for this platform: /etc/ssl/certs, /etc/pki/ca-trust/extracted/pem, /etc/pki/tls/certs, /usr/share/ca-certificates/cacert.org, /etc/ansible"}
```

It started failing after Nodesource moved to Cloudfront, using new SSL certs that use SNI (which isn't easily supported by the version of Python/other deps installed on CentOS 6.

For Ubuntu, I fixed by making the same change the official Nodesource Ansible role made: https://github.com/geerlingguy/ansible-role-nodejs/commit/0372961b152fe496412b75316a1a734b4771ad3e#diff-ebbe0c3ff3753db31a12af649c45a099R7 — see https://github.com/nodesource/ansible-nodejs-role/pull/34/files#diff-2444ad0870f91f17ca6c2a5e96b26823R16

For CentOS, not exactly sure what to do to get 6 up and running again.

**Comment by trevorr** [comment]
Assuming EPEL is installed for `python-pip`, installing the following will enable SNI for Python 2.6:

```
sudo yum install -y python-pip python-devel libffi-devel openssl-devel
sudo pip install urllib3 pyopenssl ndg-httpsclient pyasn1
```

Or in Ansible:

```
  tasks:
    - name: "Python 2.6 SNI support: install pip and required devel packages"
      yum: >
        pkg={{item}}
        state=installed
      with_items:
        - python-pip
        - python-devel
        - libffi-devel
        - openssl-devel
    - name: "Python 2.6 SNI support: install Python modules"
      pip: >
        name={{item}}
        state=present
      with_items:
        - urllib3
        - pyopenssl
        - ndg-httpsclient
        - pyasn1
```

Of course, it may break again eventually:

> Warning: Python 2.6 is no longer supported by the Python core team, please upgrade your Python. A future version of cryptography will drop support for Python 2.6

😢

**Comment by geerlingguy** [comment]
@trevorr - Yeah... at this point I'm wondering if it's worth it to put in a hack on top of a bandaid to get things working on 6; EOL isn't until 2020... but if Nodesource isn't going to reliably support a way of installing on 6, I'm not sure if I can do it easily.

Alternatively, if there's a new simpler/better way of installing Node.js via packages than Nodesource that works on RHEL/Debian flavors reliably and is well-maintained, I'm game for that.

**Comment by trevorr** [comment]
It should probably just be a separate role that is a prerequisite for this one on EL6. If there's not a way to make a role optional (for those who don't care about EL6), it could just be a documentation note that the role must be used upstream when targeting EL6.

**Comment by mlococo** [comment]
FYI, the workaround proposed by @trevorr fails with ansible 2.2.0.0 on the controller and urllib3 1.19.1 on the target. Rolling urllib3 back to 1.18.1 on the target allows the repo_key step to complete successfully on CentOS 6.

Apologies for filling the issue-db with workaround notes, but trevorr's notes allowed me to use the role without forking it, but only after the version fix.

**Comment by trevorr** [comment]
Strange, as the [urllib3 1.19.1 release notes](https://pypi.python.org/pypi/urllib3/1.19.1#changes) don't indicate any obvious changes that would break it. @mlococo, could you provide more info on the failure in that case?

**Comment by mlococo** [comment]
@trevorr I'm not sure what other details would be useful.

* Running ansible 2.2.0.0 with the geerlingguy.nodejs role acquired from galaxy a few days ago (I don't see any version in the tree that I can correlate with a git tag, but I assume it's 3.3.1).
* Running against a CentOS target in an LXD container, a slight variation on images:centos/6/amd64, which is a straightforward minimial-ish install of CentOS 6.
* I precede this role with my own role that pre-installs the SNI requirements for Python 6, exactly as you outline in your workaround. urllib3==1.19.1, pyOpenSSL 16.2.0, ndg-httpsclient==0.4.2, pyasn1==0.1.9.
* I get exactly the SNI error described in the original filing of this ticket when downloading the rpm_key of the nodejs yum repository.
* I modify my precursor recipe to specify urllib3=1.18.1 and rerun against a fresh target. nodejs install proceeds as desired.

I also inspected various changelogs and found nothing to indicate that there SHOULD be a breaking change, but given the evidence here that it worked for someone I brute-forced version combinations until I found one that worked.

**Comment by timja-kainos** [comment]
I ran into this today
I tried to work around it with:

```
# https://github.com/geerlingguy/ansible-role-nodejs/issues/46#issuecomment-247100190
yum install -y python-pip python-devel libffi-devel openssl-devel gcc
sudo pip install urllib3==1.18.1 pyopenssl==16.2.0 ndg-httpsclient==0.4.2 pyasn1==0.1.9  
```

But I'm getting:
```
==> default: fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "msg": "Failure downloading https://rpm.nodesource.com/pub_7.2.1/el/6/x86_64/nodesource-release-el6-1.noarch.rpm, 'NoneType' object has no attribute 'read'"}
```

Not quite sure where to go from here.

**Comment by geerlingguy** [comment]
This should be fixed in master/HEAD as of today, from #55 and it's follow-up. Closing, but please let me know if you still find any problems.
